### PR TITLE
Add android support

### LINKED ISSUE #310: Add Android development support
Based on https://github.com/tadfisher/android-nixpkgs

### PR DESCRIPTION
**Pull request by k3yss**
This PR adds Android support to devenv. 

This integration works with the latest channel, but creating an AVD through `android studio` is not possible (Use CLI for this). To try out this PR, you need to change the `devenv` input to `github:k3yss/work/keys/android-support?dir=src/modules`  in  your `devenv.yaml` file. 

`devenv.yaml`
```yaml
inputs:
  devenv:
    url: path: github:k3yss/work/keys/android-support?dir=src/modules
allowUnfree: true

```   

After enabling the input, you can add `android.enable=true` in your `devenv.nix` file.

`devenv.nix`

```nix
{ ... }:

{
  android.enable = true;
 
 #to use flutter we can do:
 android.flutter.enable = true;

 #to use react-native we can do:
 android.reactNative.enable = true; 
}

```
You should have an base working Android environment ready. 

Below are a few working screenshots:
 
Android Emulator working:
image

React Native working:
image

Flutter on android working: 
image

Flutter on linux  working:
image

**TODO**
- [x] Test with React Native
- [x]  Test with flutter
- [x] Add example
- [x] Make options modular
- [x] Make sure emulator works

Fixes #310

### CODE REVIEW

**Comment by k3yss** [comment]
Ready for review : )
cc @domenkozar @sandydoo

File: src/modules/integrations/android.nix [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    env.ANDROID_HOME = "${androidSdk}/libexec/android-sdk";
+    env.ANDROID_NDK_ROOT = "${config.env.ANDROID_HOME}/ndk/";
+
+    # override the aapt2 binary that gradle uses with the patched one from the sdk
+    env.GRADLE_OPTS = "-Dorg.gradle.project.android.aapt2FromMavenOverride=${androidSdk}/libexec/android-sdk/build-tools/${lib.head cfg.buildTools.version}/aapt2";
+
+    enterShell = ''
+      set -e
+      export PATH="$PATH:${config.env.ANDROID_HOME}/tools:${config.env.ANDROID_HOME}/tools/bin:${config.env.ANDROID_HOME}/platform-tools"
+      cat <<EOF > local.properties
```
Comment (domenkozar) [inline src/modules/integrations/android.nix:253]: Does this need to live in the root of the project?

File: src/modules/integrations/android.nix [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        "intel-android-sysimage-license"
+        "mips-android-sysimage-license"
+      ];
+      description = ''
+        The additional Android licenses to accept.
+        By default, several standard licenses are accepted.
+      '';
+    };
+
+    jdk = lib.mkOption {
```
Comment (domenkozar) [inline src/modules/integrations/android.nix:212]: Should we instead use languages.java.enable?

File: src/modules/integrations/android.nix [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    buildTools.version = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ "30.0.3" ];
+      description = ''
+        The version of the Android build tools to install.
+        By default, version 30.0.3 is installed.
+      '';
+    };
+
+    includeEmulator.enable = lib.mkOption {
```
Comment (domenkozar) [inline src/modules/integrations/android.nix:123]:
```suggestion
    emulator.enable = lib.mkOption {
```

File: src/modules/integrations/android.nix [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    includeSystemImages.enable = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = ''
+        Whether to include the Android system images.
+        By default, the system images are included.
+      '';
+    };
+
+    includeNDK.enable = lib.mkOption {
```
Comment (domenkozar) [inline src/modules/integrations/android.nix:159]:
```suggestion
    ndk.enable = lib.mkOption {
```

**Review by domenkozar — CHANGES_REQUESTED** [review]

**Comment by numinit** [comment]
We are about to get better AVD support with android studio: https://github.com/NixOS/nixpkgs/pull/315233

File: src/modules/integrations/android.nix (hunk 1 above)
Comment (numinit) [reply src/modules/integrations/android.nix:251]: Yes it does.

**Comment by k3yss** [comment]
Hey @numinit , thanks for the information, I have been patiently following the PR. One issue that users currently face is the creation of an emulator from `Android Studio`, I tried playing around with various env variables to come up with a potential fix for it but was unsuccessful. Did you manage to come up with a solution for it or do we expect the user to create `emulators` via `avdmanager` tool?

**Comment by numinit** [comment]
@k3yss The trick I've found is twofold: the wrapper makes new projects default their sdk.dir to the androidenv composition, but you still need the sdk.dir (and possibly ndk.dir) in your local.properties. Any new AVDs created with Android Studio work fine if this is true.

**Comment by k3yss** [comment]
@numinit What worked for me is:
1: Create a `AVD` using the CLI command.
2: Open the project in android studio.
3: Run the start command. 
- In case of `react-native` I did `npm run android` from the terminal which worked!

What I wanted to ask if it was possible to do step 1 using android studio.

**Comment by numinit** [comment]
It is possible to do step 1 in Android Studio. It's a little counterintuitive though, you need a project open with sdk.dir set properly for it to work right.

**Comment by k3yss** [comment]
@numinit I think I already have that configured, let's do a merge in the current state and wait for reviews.

**Comment by k3yss** [comment]
@domenkozar can you try the flutter configuration. I have some configuration issues with home manager.

File: docs/integrations/android.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+{ pkgs, ... }:
+
+{
+  android.enable = true;
+}
+```
+
+For a more tailored development environment, you can specify additional options:
+```nix title="env"
+env.ANDROID_HOME = "${androidSdk}/libexec/android-sdk";
```
Comment (domenkozar) [inline docs/integrations/android.md:15]: This should be overridable via `android.package`

File: docs/integrations/android.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+```nix title="devenv.yaml"
+inputs:
+  nixpkgs:
+    url: github:cachix/devenv-nixpkgs/rolling
+allowUnfree: true
+```
+## Flutter
+The following config works with the flutter starter project.
+
+```nix title="devenv.nix"
```
Comment (domenkozar) [inline docs/integrations/android.md:102]: We should hide this under `flutter.enable = true;`

**Comment by k3yss** [comment]
flutter seems to be broken, investigating.

**Comment by k3yss** [comment]
Everything works as expected for now, hope I didn't miss something 🤞

**Pull request merged**